presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EliminateCrossJoins.java

```diff
@@ -195,13 +199,23 @@ public PlanNode visitPlan(PlanNode node, RewriteContext<PlanNode> context)
                         filter);
             }
 
-            if (!graph.getAssignments().isPresent()) {
-                return result;
+            if (graph.getAssignments().isPresent()) {
+                result = new ProjectNode(
+                        idAllocator.getNextId(),
+                        result,
+                        Assignments.copyOf(graph.getAssignments().get()));
             }
-            return new ProjectNode(
-                    idAllocator.getNextId(),
-                    result,
-                    Assignments.copyOf(graph.getAssignments().get()));
+
+            if (!result.getOutputSymbols().equals(node.getOutputSymbols())) {
+                // Introduce a projection to constrain the outputs to what was originally expected
```

Comment:
maybe it would be possible to extract join node creation with enfoced output to a separate static method, e.g: `joinNodeWithEnforcedOutput`. It would add enforcing `ProjectNode` when join is cross join and therefore cannot prune outputs.